src/math/lut_trig.c

```diff
@@ -0,0 +1,108 @@
+	 45912,  46056,  46199,  46341,  46483,  46624,  46765,  46906,  47046,  47186,  47325,
+	 47464,  47603,  47741,  47878,  48015,  48152,  48288,  48424,  48559,  48694,  48828,
+	 48962,  49095,  49228,  49361,  49493,  49624,  49756,  49886,  50016,  50146,  50275,
+	 50404,  50532,  50660,  50787,  50914,  51041,  51166,  51292,  51417,  51541,  51665,
+	 51789,  51911,  52034,  52156,  52277,  52398,  52519,  52639,  52759,  52878,  52996,
+	 53114,  53232,  53349,  53465,  53581,  53697,  53812,  53926,  54040,  54154,  54267,
+	 54379,  54491,  54603,  54714,  54824,  54934,  55043,  55152,  55260,  55368,  55476,
+	 55582,  55689,  55794,  55900,  56004,  56108,  56212,  56315,  56418,  56520,  56621,
+	 56722,  56823,  56923,  57022,  57121,  57219,  57317,  57414,  57511,  57607,  57703,
+	 57798,  57892,  57986,  58079,  58172,  58265,  58356,  58448,  58538,  58628,  58718,
+	 58807,  58896,  58983,  59071,  59158,  59244,  59330,  59415,  59499,  59583,  59667,
+	 59750,  59832,  59914,  59995,  60075,  60156,  60235,  60314,  60392,  60470,  60547,
+	 60624,  60700,  60776,  60851,  60925,  60999,  61072,  61145,  61217,  61288,  61359,
+	 61429,  61499,  61568,  61637,  61705,  61772,  61839,  61906,  61971,  62036,  62101,
+	 62165,  62228,  62291,  62353,  62415,  62476,  62536,  62596,  62655,  62714,  62772,
+	 62830,  62886,  62943,  62998,  63054,  63108,  63162,  63215,  63268,  63320,  63372,
+	 63423,  63473,  63523,  63572,  63621,  63668,  63716,  63763,  63809,  63854,  63899,
+	 63944,  63987,  64031,  64073,  64115,  64156,  64197,  64237,  64277,  64316,  64354,
+	 64392,  64429,  64465,  64501,  64536,  64571,  64605,  64639,  64672,  64704,  64735,
+	 64766,  64797,  64827,  64856,  64884,  64912,  64940,  64967,  64993,  65018,  65043,
+	 65067,  65091,  65114,  65137,  65159,  65180,  65200,  65220,  65240,  65259,  65277,
+	 65294,  65311,  65328,  65343,  65358,  65373,  65387,  65400,  65413,  65425,  65436,
+	 65447,  65457,  65467,  65476,  65484,  65492,  65499,  65505,  65511,  65516,  65521,
+	 65525,  65528,  65531,  65533,  65535,  65535,  65535
+};
+
+/* Sine lookup table read */
+static inline int32_t sofm_sine_lookup_16b(int idx)
+{
+	uint16_t s;
+	int i1;
+
+	i1 = idx & (2 * SOFM_LUT_SINE_NQUART - 1);
+	if (i1 > SOFM_LUT_SINE_NQUART)
+		i1 = 2 * SOFM_LUT_SINE_NQUART - i1;
+
+	s = sofm_lut_sine_table_s16[i1];
+	if (idx > 2 * SOFM_LUT_SINE_NQUART)
+		return -((int32_t)s);
+
+	return (int32_t)s;
+}
+
+/* Compute fixed point sine with table lookup and interpolation */
+int16_t sofm_lut_sin_fixed_16b(int32_t w)
+{
+	int64_t idx;
+	int32_t sine;
+	int32_t frac;
+	int32_t delta;
+	int32_t s0;
+	int32_t s1;
+	int64_t idx_tmp;
+
+	/* Q4.28 x Q12.20 -> Q16.48 --> Q16.31*/
+	idx_tmp = ((int64_t)w * SOFM_LUT_SINE_C_Q20) >> 17;
+	idx = (idx_tmp >> 31); /* Shift to Q0 */
+	frac = (int32_t)(idx_tmp - (idx << 31)); /* Get fraction Q1.31*/
```

Comment:
that seems to boil down to
```
idx_tmp - (idx_tmp & 0xffffffff80000000ULL) ==
idx_tmp & 0x7fffffff
```
would the compiler optimise that out by itself?

Reply:
I was thinking that but it looks awkward in arithmetic that is not bit-banging to HW registers etc. The shifts have association to Qx.y format. But if that gives MCPS advantage I can change, and comment what happens, I'll try.

Reply:
With this modification 69.976 to 69.925 MCPS, not worth it I think, because of bit-and awkwardness here. I think our perf measurement works down to 0.1 MCPS level, below that it's probably noise.